### PR TITLE
Share block cache among column families

### PR DESCRIPTION
**Pull request by yiwu-arbug**
## What have you changed? (mandatory)
Add option to share block cache between column families. The goal is to simplify config. It may also improve cache performance if previous config fail to balance cache size well. A previous benchmark indeed show read throughput and latency increases in the specific test scenario (see benchmark result link).

The new config provide a flag to enable sharing block cache, and it is default true. Cache capacity is optional. If it is not provided, it is decided by 
* the sum of cache size of individual column families. and
* In the particular case when cache sizes for all column families are not set, the shared cache size is ~45% of total physical memory.

We also expose more block cache options through the new block cache config.

Config example:
```
[storage.block-cache]
shared = true
capacity = "10G"
num-shard-bits = -1
strict-capacity-limit = false
high-pri-pool-ratio = 0.0
```

## What are the type of the changes? (mandatory)
New feature

## How has this PR been tested? (mandatory)
Run tikv-server from command line. Change tikv config with the following scenario. Check the cache address and detailed cache options in rocksdb log.
* case 1: shared=false. Expect: block cache is not shared, and their individual sizes respect individual block-cache-size setting.
* case 2: shared=true and capacity is not provided. block cache size is provided for individual column families. Expect: Cache is shared and with size equal to sum of individual cache sizes.
* case 3: shared=true and capacity is not provided. block cache size is not provided for individual column families. Expect: Cache is shared and with size ~=45% total physical memory.
* case 4: shared=true and capacity is provided. Expect: cache is shared and with size equal to capacity.

## Does this PR affect documentation (docs) update? (mandatory)
Yes. We will need to update our tuning guide to suggest users to set shared cache instead.

## Does this PR affect tidb-ansible update? (mandatory)
Yes. We need to update config template.

## Benchmark result if necessary (optional)
Test with The following YCSB with steps against TiKV transaction API.

Load 30M records into empty TiKV instance, with workload A.
Run workload C (point lookup only) with 30M operations, from 500 threads.
Run workload E (95% small scan + 5% updates) with 3M operations, from 500 threads.
The resulting DB size is ~36G. 

Details test script:

./bin/go-ycsb load tikv -P workloads/workloada -p tikv.pd="172.16.30.40:2379" -p tikv.type="txn" -p recordcount=30000000
./bin/go-ycsb run tikv -P workloads/workloadc -p tikv.pd="172.16.30.40:2379" -p tikv.type="txn" -p recordcount=30000000 -p threadcount=500 -p operationcount=30000000

Testing the following versions:
Master: default CF 10G, lock CF 1G, write CF 10G, Raft CF 128M, Raft RocksDB: 1G
Test 1: KV RocksDB: shared block cache with total size 21G. Raft RocksDB: 1G
Test 2: KV RocksDB: shared block cache with total size 14G. Raft RocksDB: 1G

Result: 
Master (block cache hit rate: 95.4%)
READ - Takes(s): 442.3, Count: 29999994, OPS: 67821.4, Avg(us): 7366, Min(us): 326, Max(us): 91756, 95th(us): 9000, 99th(us): 11000

Test 1 (throughput +16%, average latency -14%, block cache hit rate: 97.7%):

READ - Takes(s): 380.1, Count: 29999994, OPS: 78930.5, Avg(us): 6329, Min(us): 311, Max(us): 90951, 95th(us): 8000, 99th(us): 11000

Test 2 (throughput +4.6%, average latency -4.4%, block cache hit rate: 96.1%):

READ - Takes(s): 422.8, Count: 29999999, OPS: 70958.6, Avg(us): 7040, Min(us): 257, Max(us): 85152, 95th(us): 9000, 99th(us): 11000

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
@zhangjinpeng1987 @Connor1996 Mind take a look if the config structure is good? If so, I'll continue to update config template and add unit tests. Thanks.

**Comment by zhangjinpeng1987** [comment]
Mostly LGTM, please don't forget to modify etc/config-template.toml.

**Comment by yiwu-arbug** [comment]
@zhangjinpeng1987 @Connor1996 updated config template and added unit test. mind take a look again?

Thing not done in this patch and I'll work on later (soon):
* update grafana to show shared cache capacity. I didn't test but I believe right now it will show multiple cache sizes with the same value, and it will be confusing.
* update tikv-ctl to allow dynamic change of the shared cache size. I didn't test but I believe right now we can achieve the same by dynamic setting cache size of an arbitrary CF, but its confusing.

File: etc/config-template.toml [hunk 1]
```diff
@@ -134,6 +134,45 @@
...
+
+## Whether to fail read when the cache reaches its capacity, but no cache items can be evicted.
+# strict-capacity-limit = false
+
+## Config ratio of cache capacity allocate to the high-pri pool. Should be a value between 0.0-1.0.
+## When set to value greater than zero, mid-point insertion strategy is enabled. Cache items will
+## first inserted to the middle of LRU list at the position decided by the high-pri pool ratio,
+## and move to the tail of LRU list after at least one cache hit. Enabling mid-point insertion
+## will provide some resistence to large table scan evicting useful cache items.
+# high-pri-pool-ratio = 0.0
```
Comment (yiwu-arbug) [inline etc/config-template.toml]: maybe we shouldn't even mention `num-shard-bits`, `strict-capacity-limit` and `high-pri-pool-ratio` in the template?

Comment (zhangjinpeng1987) [reply etc/config-template.toml]: Good idea, less is more.

File: tests/integrations/config/test-custom.toml [hunk 2]
```diff
@@ -59,6 +59,13 @@ scheduler-concurrency = 123
 scheduler-worker-pool-size = 1
 scheduler-pending-write-threshold = "123KB"
 
+[storage.block-cache]
+shared = true
+capacity = "40GB"
+num-shard-bits = 6
```
Comment (zhangjinpeng1987) [inline tests/integrations/config/test-custom.toml]: For test, i thint not default valule is better.

**Comment by siddontang** [comment]
Thanks @yiwu-arbug 

But the confluence can't be accessed outside, please paste the benchmark result directly here.

File: etc/config-template.toml [hunk 3]
```diff
@@ -681,7 +690,6 @@
 
 ## Generally, you can set it from 256MB to 2GB. In most cases, you can use the default value. But
```
Comment (siddontang) [inline etc/config-template.toml]: you should remove the comments too

File: etc/config-template.toml [hunk 4]
```diff
@@ -134,6 +134,30 @@
 ## When the pending write bytes exceeds this threshold, the "scheduler too busy" error is displayed.
 # scheduler-pending-write-threshold = "100MB"
 
+[storage.block-cache]
```
Comment (siddontang) [inline etc/config-template.toml]: have you missed other configurations for the shared cache in the config file?

Comment (yiwu-arbug) [reply etc/config-template.toml]: Discussed with @zhangjinpeng1987 and we think normal users may not need to change them and we should make them hidden. https://github.com/tikv/tikv/pull/4563#discussion_r278748010

**Comment by yiwu-arbug** [comment]
@zhangjinpeng1987 @Connor1996 mind take another look? I'll leave grafana metrics and tikv-ctl changes to another PR. Thanks.

**Comment by yiwu-arbug** [comment]
@siddontang @Connor1996 mind give an approval? Thanks!

**Pull request merged**